### PR TITLE
Support websocket reconnection when the server status is abnormal

### LINKED ISSUE #162: [Federated Learning] train worker was error when it was failed firstly to connect the avg worker
**What happened**:
Following [this federated learning example](), I created a federated learning job.
Train worker failed firstly to connect avg worker,  because the avg worker didn't run successfully.
Then, train worker was error and restarted.
Below is the error log of the train worker:

```
2021-08-17 08:01:21.380224: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
[2021-08-17 08:01:23,885] federated_learning.py(57) [INFO] - Node trainworker-lgszc connect to : ws://172.19.0.2:32254/FedAvg
[2021-08-17 08:01:23,886] client.py(153) [INFO] - ws://172.19.0.2:32254/FedAvg connection by trainworker-lgszc
[2021-08-17 08:01:24,898] client.py(165) [INFO] - ws://172.19.0.2:32254/FedAvg connection was refused by server
Traceback (most recent call last):
  File "train.py", line 81, in <module>
    main()
  File "train.py", line 67, in main
    aggregation=aggregation_algorithm)
  File "/home/lib/sedna/core/federated_learning/federated_learning.py", line 53, in __init__
    self.register(timeout=connect_timeout)
  File "/home/lib/sedna/core/federated_learning/federated_learning.py", line 61, in register
    ping_timeout=timeout
  File "/home/lib/sedna/service/client.py", line 149, in __init__
    asyncio.wait_for(self.connect(), timeout=timeout)
  File "/usr/lib/python3.6/asyncio/base_events.py", line 484, in run_until_complete
    return future.result()
  File "/usr/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/home/lib/sedna/service/client.py", line 158, in connect
    ), self._ws_timeout)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/lib/python3.6/asyncio/tasks.py", line 537, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/usr/local/lib/python3.6/dist-packages/websockets/legacy/client.py", line 622, in __await_impl__
    transport, protocol = await self._create_connection()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 794, in create_connection
    raise exceptions[0]
  File "/usr/lib/python3.6/asyncio/base_events.py", line 781, in create_connection
    yield from self.sock_connect(sock, address)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 439, in sock_connect
    return (yield from fut)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 469, in _sock_connect_cb
    raise OSError(err, 'Connect call failed %s' % (address,))
ConnectionRefusedError: [Errno 111] Connect call failed ('172.19.0.2', 32254)
```
**What you expected to happen**:
When train worker cannot connect to the avg worker, it should try N(e.g., 5) times.
After N times's tries,  the train worker should be restarted.

**Why is this needed**:
1. the avg worker restarts unexpectedly
2. edge node offline and the network is abnormal
3. GM can not create train worker and avg worker almost at the same time. 

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
<details><summary>Sedna Version: v0.3.1</summary>

```console
$ kubectl get -n sedna deploy gm -o jsonpath='{.spec.template.spec.containers[0].image}'
# paste output here

$ kubectl get -n sedna ds lc -o jsonpath='{.spec.template.spec.containers[0].image}'
# paste output here
```

</details>

<details><summary>Kubernets Version</summary>

```console
$ kubectl version
# paste output here
```

</details>

<details><summary>KubeEdge Version</summary>

```console
$ cloudcore --version
# paste output here

$ edgecore --version
# paste output here
```

</details>

**CloudSide Environment**:
<details><summary>Hardware configuration</summary>

```console
$ lscpu
# paste output here
```

</details>

<details><summary>OS</summary>

```console
$ cat /etc/os-release
# paste output here
```

</details>

<details><summary>Kernel</summary>

```console
$ uname -a
# paste output here
```

</details>

<details><summary>Others</summary>
</details>

**EdgeSide Environment**:
<details><summary>Hardware configuration</summary>

```console
$ lscpu
# paste output here
```

</details>

<details><summary>OS</summary>

```console
$ cat /etc/os-release
# paste output here
```

</details>

<details><summary>Kernel</summary>

```console
$ uname -a
# paste output here
```

</details>

<details><summary>Others</summary>
</details>

### PR DESCRIPTION
**Pull request by JoeyHwong-gk**
fix #162 
Signed-off-by: JoeyHwong <joeyhwong@gknow.cn>

### CODE REVIEW

File: lib/sedna/service/client.py [hunk 1]
```diff
@@ -151,39 +151,34 @@ def __init__(self, url, client_id, **kwargs):
...
+                LOGGER.warning(f"{self.uri} connection closed")
+            except InvalidStatusCode as err:
+                LOGGER.warning(
+                    f"{self.uri} websocket failed - "
+                    f"with invalid status code {err.status_code}")
+            except WebSocketException as err:
+                LOGGER.warning(f"{self.uri} websocket failed - with {err}")
+            except OSError as err:
+                LOGGER.warning(f"{self.uri} connection failed - with {err}")
+            except Exception:
```
Comment (llhuii) [inline lib/sedna/service/client.py:178]: better to not capture **all** exceptions, you can just raise it if unexpected.

File: lib/sedna/service/client.py [hunk 2]
```diff
@@ -151,39 +151,34 @@ def __init__(self, url, client_id, **kwargs):
...
-        except WebSocketException as err:
-            LOGGER.info(f"{self.uri} websocket failed - with {err}")
-            raise
-        except OSError as err:
-            LOGGER.info(f"{self.uri} connection failed - with {err}")
-            raise
-        except Exception:
-            LOGGER.exception(f"{self.uri} websocket Error")
-            raise
+        for _ in range(self._retry):
```
Comment (llhuii) [inline lib/sedna/service/client.py:154]: I note you have already use `retry` package, you can use it here again.

Comment (JoeyHwong-gk) [reply lib/sedna/service/client.py]: fixed. As `retry` no longer maintained, I'll replace it with `tenacity`

File: lib/sedna/service/client.py (hunk 1 above)
Comment (JoeyHwong-gk) [reply lib/sedna/service/client.py]: check

**Pull request merged**